pageserver/src/relish_storage/storage_sync.rs

```diff
@@ -0,0 +1,1416 @@
+    max_concurrent_sync: usize,
+    relish_storage: &'a S,
+    layers: impl Iterator<Item = PathBuf>,
+    metadata_path: &'a Path,
+    sync_operation: SyncOperation,
+) -> SyncResult {
+    let concurrent_operations_limit = Arc::new(Semaphore::new(max_concurrent_sync));
+    let mut sync_operations = layers
+        .into_iter()
+        .map(|layer_path| {
+            let limit = Arc::clone(&concurrent_operations_limit);
+            async move {
+                let permit = limit.acquire().await.expect("Semaphore is not closed yet");
+                let sync_result = match sync_operation {
+                    SyncOperation::Download => download(relish_storage, &layer_path).await,
+                    SyncOperation::Upload => upload(relish_storage, &layer_path).await,
+                };
+                drop(permit);
+                (layer_path, sync_result)
+            }
+        })
+        .collect::<FuturesUnordered<_>>();
+
+    let mut synced = Vec::new();
+    let mut not_synced = Vec::new();
+    while let Some((layer_path, layer_download_result)) = sync_operations.next().await {
+        match layer_download_result {
+            Ok(()) => synced.push(layer_path),
+            Err(e) => {
+                error!(
+                    "Failed to sync ({:?}) layer with local path '{}', reason: {:#}",
+                    sync_operation,
+                    layer_path.display(),
+                    e,
+                );
+                not_synced.push(layer_path);
+            }
+        }
+    }
+    concurrent_operations_limit.close();
+
+    if not_synced.is_empty() {
+        debug!(
+            "Successfully synced ({:?}) all {} layers",
+            sync_operation,
+            synced.len(),
+        );
+        trace!("Synced layers: {:?}", synced);
+        let metadata_sync_result = match sync_operation {
+            SyncOperation::Download => download(relish_storage, metadata_path).await,
+            SyncOperation::Upload => upload(relish_storage, metadata_path).await,
+        };
+        match metadata_sync_result {
+            Ok(()) => {
+                debug!("Metadata file synced successfully");
+                SyncResult::Success { synced }
+            }
+            Err(e) => {
+                error!(
+                    "Failed to sync ({:?} metadata file with local path '{}', reason: {:#}",
+                    sync_operation,
+                    metadata_path.display(),
+                    e
+                );
+                SyncResult::MetadataSyncError { synced }
+            }
+        }
+    } else {
+        SyncResult::LayerSyncError { synced, not_synced }
+    }
+}
+
+async fn upload<P, S: 'static + RelishStorage<RelishStoragePath = P>>(
+    relish_storage: &S,
+    source: &Path,
+) -> anyhow::Result<()> {
+    let destination = relish_storage.storage_path(source).with_context(|| {
+        format!(
+            "Failed to derive storage destination out of metadata path {}",
```

Comment:
Why does this message mention `metadata`? I assume we may upload any type of file here.

Reply:
Wrong, fixed, thanks.